src/objective/multiclass_obj_gpu.cu

```diff
@@ -0,0 +1,245 @@
+
+    max_data_.Reshard(devices_);
+  };
+
+  void GetGradient(HostDeviceVector<bst_float> *preds, const MetaInfo &info,
+                   int iter,
+                   HostDeviceVector<GradientPair> *out_gpair) override {
+    CHECK_NE(info.labels_.size(), 0U) << "label set cannot be empty";
+    CHECK(preds->Size() ==
+          (static_cast<size_t>(param_.num_class) * info.labels_.size()))
+        << "SoftmaxMultiClassObj: label size and pred size does not match";
+
+    const int nclass = param_.num_class;
+    size_t n = preds->Size();
+    auto ndata = static_cast<size_t>(n / nclass);  // number of data points
+
+    preds->Reshard(devices_);
+    out_gpair->Reshard(devices_);
+    out_gpair->Resize(preds->Size());
+
+    LazyResize(info.labels_.size(), info.weights_.size());
+
+    if (!copied_) {
+      labels_.Copy(info.labels_);
+      if (info.weights_.size() > 0) {
+        weights_.Copy(info.weights_);
+      } else {
+        copied_ = true;
+      }
+    }
+
+    int h_label_error = 0;
+    int *d_label_error;
+    cudaMalloc(&d_label_error, sizeof(int));
+#pragma omp parallel for schedule(static, 1) if (devices_.Size() > 1)
+    for (int i = 0; i < devices_.Size(); ++i) {
+      int d = devices_[i];
+      dh::safe_cuda(cudaSetDevice(d));
+
+      const int block = 256;
+
+      multiclass_gradient_k<<<dh::DivRoundUp(ndata, block), block>>>
+        (out_gpair->DevicePointer(d), labels_.DevicePointer(d),
+         preds->DevicePointer(d),
+         info.weights_.size() > 0 ? weights_.DevicePointer(d) : nullptr,
+         ndata, nclass, d_label_error);
+      dh::safe_cuda(cudaGetLastError());
+    }
+    dh::safe_cuda(cudaDeviceSynchronize());
+    cudaMemcpy(&h_label_error, d_label_error, sizeof(int),
+               cudaMemcpyDeviceToHost);
+    cudaFree(d_label_error);
+
+    CHECK(h_label_error >= 0 && h_label_error < nclass)
+        << "SoftmaxMultiClassObj: label must be in [0, num_class),"
+        << " num_class=" << nclass << " but found " << h_label_error
+        << " in label.";
+  }
+
+  void PredTransform(HostDeviceVector<bst_float> *io_preds) override {
+    this->Transform(io_preds, output_prob_);
+  }
+
+  void EvalTransform(HostDeviceVector<bst_float> *io_preds) override {
+    this->Transform(io_preds, true);
+  }
+
+  const char *DefaultEvalMetric() const override { return "merror"; }
+
+ private:
+  inline void Transform(HostDeviceVector<bst_float> *preds, bool prob) {
+    const int nclass = param_.num_class;
+
+    const auto ndata = static_cast<int>(preds->Size() / nclass);
+    preds->Reshard(devices_);
+    if (!prob)
+      max_data_.Resize(ndata);
+
+#pragma omp parallel for schedule(static, 1) if (devices_.Size() > 1)
+    for (int i = 0; i < devices_.Size(); ++i) {
+      int d = devices_[i];
+      dh::safe_cuda(cudaSetDevice(d));
+      const int block = 256;
+      size_t n = preds->DeviceSize(d);
+      size_t ndata = n / nclass;
+
```

Comment:
These two smaller kernels could be written in very few lines using the dh::launch_n() primitive. Search the code base for examples on how to use this.